### ISSUE TITLE
Support JSON builder on non-contiguous buffers

### ISSUE THREAD

**Issue by gilbertw** [Azure/azure-sdk-for-c#509]
First, in Azure Core, my team defines the following 2 data types:

```
typedef struct
{
// The commented-out fields below are fields WE COULD have if necessary but I suggest we do NOT
// have them now. So, this structure has just 1 field in it now (user_context). But, encapsulating
// this field in this allocator_context struct allows us to add fields to this struct in the
// future without breaking customer code.

// int remaining_length;
// int bytesUsedInlastSpan;
// int numberSpansUsedSoFar;
void *user_context;
} allocator_context;


// This data type is the signature of a callback function. The caller of az_span_append_chunked implements this function
// to allocate a new span and return it back to az_span_append_chunked so it can append to a dis-contiguous chunk:
typedef az_result (*span_allocator_fn)(allocator_context *allocator_context, az_span *out_next_destination);
```

Second, in Azure Core, my team defines the following function which appends a ‘source’ span to a set of dis-contiguous spans. The first span to use is passed in ‘first_destination’. If this span is big enough to hold the source, then the source is copied into first_destination and the function returns FAST. The customer knows how many bytes are copied to dis-contiguous spans by simply looking at the length of the source span. If the caller knows that first_destination is big enough to hold the source, the caller can pass NULL for both the allocator_callback and user_context parameters. The out_destiantion_remainder is populated by the function with a span reflecting the remainder of the last dis-contiguous span used. The caller can use the remainder of this span if they desire. Typically, it will be used to call az_span_append_chunked again to append an additional source.

```
az_result az_span_append_chunked(az_span first_destination, az_span source, span_allocator_fn allocator_callback, void *user_context, az_span *out_destination_remainder)
{
az_span destination = first_destination;
uint8_t *src_ptr = az_span_ptr(source); // Start of source to copy
int32_t src_len = az_span_length(source); // Bytes left of source to copy

while (1)
{
    // Write into destination as much source as will fit
    int32_t dst_remainder = az_span_capacity(destination) - az_span_length(destination);
    int32_t copy_len = (src_len < dst_remainder) ? src_len : dst_remainder; // Copy which ever is smaller
    memcpy(&(az_span_ptr(destination)[az_span_length(destination)]), src_ptr, copy_len);
    if (copy_len == src_len)
    {
        *out_destination_remainder = az_span_set_length(first_destination, az_span_length(destination) + copy_len);
        return AZ_OK; // We successfully copied all of the source, return success
    }
    src_ptr += copy_len; // Position source past the bytes we've already copied
    src_len -= copy_len; // The remaining source length after copying some bytes

    // The source has more bytes which didn't fit in the destination; call allocator to get a new az_span
    allocator_context ac = {.user_context = user_context, .remaining_length = src_len - copy_len};
    az_result r = (allocator_callback == NULL) ? 0 /* TODO: AZ_SOURCE_TOO_BIG*/ : allocator_callback(&ac, &destination);
    if (r != AZ_OK)
       return r; // if allocator fails, return error to caller; else continue with new destination span
}
}
```

Third, the IOT team implements a function like this (what I show below is abbreviated just to show an example). The caller of this (example: Azure RTOS Convenience layer) passes the first_destination span, an allocator and an allocator_content.
Internally, this function can call az_span_append_chunked multiple times to append multiple sources (the example below appends a username and then a password); the function then returns the total length written; we might also want the remaining span here (I’m not sure).

```
az_result pnp_client_get_topic_chunked(az_span first_destination, span_allocator_fn allocator, void *allocator_context, int32_t *out_length)
{
az_span destination = first_destination;
az_span username, password;
az_span_append_chunk(destination, username, allocator, allocator_context, &destination);
az_span_append_chunk(destination, password, allocator, allocator_context, &destination);
*out_length = az_span_length(username) + az_span_length(password);
return AZ_OK;
}
```

Fourth, the IOT team ALSO implements a function like this (again abbreviated). This function is for the non-RTOS customers that have a single contiguous span; it hides from them the chunked/allocator stuff (everything above). Internally, this function is a thin wrapper over the chunked functionality. Implementing the functions this way means that the IOT team can do all the “important work” in the chunked function just once and this is where bugs/features will be addressed. The thin/simple wrapper code just calls the complete implementation.

```
az_result pnp_client_get_topic(az_span topic, az_span *out_topic)
{
int32_t length;
az_result r = pnp_client_get_topic_chunked(topic, NULL, NULL, &length);
*out_topic = az_span_set_length(topic, length);
return r;
}
```

**Comment by gilbertw** [comment]
@JeffreyRichter please add the design requirements and design proposal to the description

**Comment by JeffreyRichter** [comment]
I’ve come up with a design and reviewed it Yuxin and Ahson:
First, in Azure Core, my team defines the following 2 data types:
 
typedef struct
{
    // The commented-out fields below are fields WE COULD have if necessary but I suggest we do NOT
    // have them now. So, this structure has just 1 field in it now (user_context). But, encapsulating 
    // this field in this allocator_context struct allows us to add fields to this struct in the 
    // future without breaking customer code.
 
    // int remaining_length;
    // int bytesUsedInlastSpan;
    // int numberSpansUsedSoFar;
    void *user_context;
} allocator_context;
 
// This data type is the signature of a callback function. The caller of az_span_append_chunked implements this function
// to allocate a new span and return it back to az_span_append_chunked so it can append to a dis-contiguous chunk:
typedef az_result (*span_allocator_fn)(allocator_context *allocator_context, az_span *out_next_destination);
 
 
Second, in Azure Core, my team defines the following function which appends a ‘source’ span to a set of dis-contiguous spans. The first span to use is passed in ‘first_destination’. If this span is big enough to hold the source, then the source is copied into first_destination and the function returns FAST. The customer knows how many bytes are copied to dis-contiguous spans by simply looking at the length of the source span. If the caller knows that first_destination is big enough to hold the source, the caller can pass NULL for both the allocator_callback and user_context parameters. The out_destiantion_remainder is populated by the function with a span reflecting the remainder of the last dis-contiguous span used. The caller can use the remainder of this span if they desire. Typically, it will be used to call az_span_append_chunked again to append an additional source.
 
az_result az_span_append_chunked(az_span first_destination, az_span source, span_allocator_fn allocator_callback, void *user_context, az_span *out_destination_remainder)
{
    az_span destination = first_destination;
    uint8_t *src_ptr = az_span_ptr(source);   // Start of source to copy
    int32_t src_len = az_span_length(source); // Bytes left of source to copy
 
    while (1)
    {
        // Write into destination as much source as will fit
        int32_t dst_remainder = az_span_capacity(destination) - az_span_length(destination);
        int32_t copy_len = (src_len < dst_remainder) ? src_len : dst_remainder; // Copy which ever is smaller
        memcpy(&(az_span_ptr(destination)[az_span_length(destination)]), src_ptr, copy_len);
        if (copy_len == src_len)
        {
            *out_destination_remainder = az_span_set_length(first_destination, az_span_length(destination) + copy_len);
            return AZ_OK; // We successfully copied all of the source, return success
        }
        src_ptr += copy_len; // Position source past the bytes we've already copied
        src_len -= copy_len; // The remaining source length after copying some bytes
 
        // The source has more bytes which didn't fit in the destination; call allocator to get a new az_span
        allocator_context ac = {.user_context = user_context, .remaining_length = src_len - copy_len};
        az_result r = (allocator_callback == NULL) ? 0 /* TODO: AZ_SOURCE_TOO_BIG*/ : allocator_callback(&ac, &destination);
        if (r != AZ_OK)
           return r; // if allocator fails, return error to caller; else continue with new destination span
    }
}
 
 
Third, the IOT team implements a function like this (what I show below is abbreviated just to show an example). The caller of this (example: Azure RTOS Convenience layer) passes the first_destination span, an allocator and an allocator_content.
Internally, this function can call az_span_append_chunked multiple times to append multiple sources (the example below appends a username and then a password); the function then returns the total length written; we might also want the remaining span here (I’m not sure). 
 
az_result pnp_client_get_topic_chunked(az_span first_destination, span_allocator_fn allocator, void *allocator_context, int32_t *out_length)
{
    az_span destination = first_destination;
    az_span username, password;
    az_span_append_chunk(destination, username, allocator, allocator_context, &destination);
    az_span_append_chunk(destination, password, allocator, allocator_context, &destination);
    *out_length = az_span_length(username) + az_span_length(password);
    return AZ_OK;
}
 
 
Fourth, the IOT team ALSO implements a function like this (again abbreviated). This function is for the non-RTOS customers that have a single contiguous span; it hides from them the chunked/allocator stuff (everything above). Internally, this function is a thin wrapper over the chunked functionality. Implementing the functions this way means that the IOT team can do all the “important work” in the chunked function just once and this is where bugs/features will be addressed. The thin/simple wrapper code just calls the complete implementation. 
 
az_result pnp_client_get_topic(az_span topic, az_span *out_topic)
{
    int32_t length;
    az_result r = pnp_client_get_topic_chunked(topic, NULL, NULL, &length);
    *out_topic = az_span_set_length(topic, length);
    return r;
}
 
 
To sum up: I think this is a very sustainable approach to solving the non-contiguous buffer problem without affecting our current span implementation. This approach also doesn’t affect/impact customers who are not using an RTOS. 
My team will add this functionality to our JSON builder.
